### 목표
* json 파일 읽을수 있다.
* nosql 파일저장소 (mongo db)

In [6]:
!curl http://httpbin.org/encoding/utf8

<h1>Unicode Demo</h1>

<p>Taken from <a
href="http://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-demo.txt">http://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-demo.txt</a></p>

<pre>

UTF-8 encoded sample plain-text file
�얄��얄��얄��얄��얄��얄��얄��얄��얄��얄��얄��얄��얄��얄��얄��얄��얄��얄�

Markus Kuhn [�ma稼k�s ku�n] <http://www.cl.cam.ac.uk/~mgk25/> ��2002-07-25


The ASCII compatible UTF-8 encoding used in this plain-text file
is defined in Unicode, ISO 10646-1, and RFC 2279.


Using Unicode/UTF-8, you can write in emails and source code things such as

Mathematics and sciences:

  ��E�꿥a = Q,  n ���� ��f(i) = ��g(i),      �㎮렊�쎻뵆��������뵍�왿렎��
                                            �も렋�쒋봻a짼+b쨀 �잂렏��
  ���댿꽍: �늵��= �믠뙄�뭯�� 慣 ��짭棺 = 짭(짭慣 ��棺),    �も렋�쒋봻���������잂렏��
                                            �も렋�쒋렩 c��  �잂렏��
  �����뺚� ����������������                   �ⓥ렋��      �잂렏��
                                            �も렋����    �잂렏��
  ��< a ��b ��c ��d ������(�쪨�����찦��,      �も렋����

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 14239  100 14239    0     0  28478      0 --:--:-- --:--:-- --:--:-- 28478


In [7]:
import requests
r = requests.get("http://httpbin.org/encoding/utf8")
r.text[0:500]

u'<h1>Unicode Demo</h1>\n\n<p>Taken from <a\nhref="http://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-demo.txt">http://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-demo.txt</a></p>\n\n<pre>\n\nUTF-8 encoded sample plain-text file\n\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\u203e\n\nMarkus Kuhn [\u02c8ma\u02b3k\u028as ku\u02d0n] <http://www.cl.cam.ac.uk/~mgk25/> \u2014 2002-07-25\n\n\nThe ASCII compatible UTF-8 encoding used in this plain-text file\nis defined in Unicode, ISO 10646-1, and RFC 2279.\n\n\nUsing Unicode/UTF-8, you can write in emails and so'

* '\u' = unicode가 뒤에따라온다.
* 명령창에서는 unicode를 인식을 못하니까 저렇게 위와같이 이상하게 나타나는것이다.
* 화면에서 그 unicode값을 디코딩해서 제대로 출력하면 된다. 
* 화면에서 제대로 출력하기위한 조건 2가지
    * 

In [9]:
import locale
locale.getdefaultlocale()
#cp949 를 UTF-8로바꿔줘야 문제가 없다. CP949는 내 운영체제 자체의 인코딩 방식임.

('ko_KR', 'cp949')

* 명령창에 systeminfo라고 치면 나의 locale 정보를 알수 있다. window 10은 'ko;한국어'다. 이걸 UTF-8로 바꿔줘야함 (운영체제자체를 변경하는것은 교수님 안알려주심. 우리는 python창에서의 인코딩방식을 바꿔서 python화면에서 볼때 제대로 보게 할 거임)

* 운영체제가 그냥 UFT-8이면 python에서도 그냥 잘 될 가능성이 높다.
* python3에서는 인코딩방식을 자동으로 지원하기때문에 아래와 같은 일을 하지 않아도 될 확률이 높다. 하지만 우리는 python2를 사용하고있기때문에 아래와같은 공작을 한다.

In [11]:
import sys, locale
print locale.getpreferredencoding()
print sys.getdefaultencoding()
print sys.getfilesystemencoding()
print sys.stdin.encoding
print sys.stdout.encoding

cp949
ascii
mbcs
None
UTF-8


In [40]:
import locale
locale.setlocale(locale.LC_ALL, 'en_US.utf-8')

Error: unsupported locale setting

In [35]:
#python 인코딩방식으로 utf 8로 변경했다.
import sys
reload(sys)
sys.setdefaultencoding('UTF-8')

In [20]:
str('한')#16진법으로 표현 ( x = 16 )

'\xed\x95\x9c'

In [19]:
u'한' #unicode식으로 출력

u'\ud55c'

In [36]:
'한'.encode('utf-8') 
#에러뜨는 이유:
#그냥 한이라고 하면 처음부터 128개까지의 문자만 찾는다. 근데 그안에는 영문자 대소문자와 숫자기호만 있다.
#에러가 안뜨는데?

'\xed\x95\x9c'

In [26]:
(u'한').encode('utf-8')

'\xed\x95\x9c'

In [1]:
print u'한' #왜 아무것도 안뜨지?

한


In [2]:
st='한'
print st #얘도 아무것도 안뜨네?

한


## json파일

* schema : 행과열을 정하는것 (ex: 시간표, xml)
* schema가 없는것 : 엑셀(특정한 것이 아니고 광범위하게 행과 열이 존재하기때문)
* json은 schema가 없다

In [11]:
import json

user1={"id": "405", "name": "js"}
user2={ "id": "405", "이름": "js" } 
#16*16 = 256 = 1바이트 
#6개가 나오는이유 : 초성,중성,종성 한글자당 3개씩 나옴
user3={ "id": "405", u"이름": "js" }

print user1
print user2
print user3

{'id': '405', 'name': 'js'}
{'id': '405', '\xec\x9d\xb4\xeb\xa6\x84': 'js'}
{u'\uc774\ub984': 'js', 'id': '405'}


### json array

In [12]:
import json

p={"Persons":[
    {"id":"405", "name":"js1"},
    {"id":"406", "name":"js2"},
]}

print p

{'Persons': [{'id': '405', 'name': 'js1'}, {'id': '406', 'name': 'js2'}]}


### writing json file

* json을 파일에서 읽을때 그것은 문자열이다. 그게 json인지 그냥 문자열인지 인지가 안됨

In [15]:
import json

p={"Persons":[
    {"id":"405", "이름":"js1"},
    {"id":"406", "이름":"js2"},
]}

#json array를 문자열로 쓴다
#\u : 아 utf (unicode)로 인코딩이 되있구나 괜찮군. 출력할때만 한글로 보이기만 하면 됨.
print json.dumps(p) 

{"Persons": [{"id": "405", "\uc774\ub984": "js1"}, {"id": "406", "\uc774\ub984": "js2"}]}


In [4]:
%%writefile data/ds_save_persons.json
p={"Persons":[
    {"id":"405", "이름":"js1"},
    {"id":"406", "이름":"js2"},
]}

Writing data/ds_save_persons.json


In [5]:
import os, io, json
with io.open(os.path.join('data','ds_save_persons.json'), 'w') as json_file:
    json.dump(p, json_file, encoding="utf-8")
    
#with : python 명령어 : 파일을 열때 with로 열수 있다. with은 어떤 설정을해서 설정을 가지고 파일을 연다. 
#with 로 열면 close를 해주지 않아도 된다.


NameError: name 'p' is not defined

* mongod (mongo demon) : 서버 (아나콘다)
* mongo : 클라이언트 (명령프롬프트)
* mongodb download 를 구글창에친다. 
* 2번째꺼에 들어간다. https://www.mongodb.com/download-center#community
* 첫번째꺼를 선택해서 다운
* 다운 다되면 환경변수 설정을 해준다. 
* 시스템변수의 path의 맨끝에 C:\Program Files\MongoDB\Server\3.4\bin 이거 삽입
* 루트파일에 mymongodb 디렉토리 생성
* 아나콘다에서 루트로 가서 s_201511138에서  mongod --dbpath mymongodb 를 친다 - 패스설정

* 맨마지막에 커서가 깜박이면 서버가 생성된것이다.

#### cmd에서 몽고 쓰기

* cmd 를 열어서 몽고를 써보자 - png 참고
* s_201511138로 가서 mongo를 친다.

* use test : test라는 db가 없으면 만들고 있으면 있는거 가져다가 쓴다.
* db.testCollection.insert({id:1}) : testCollection이라는 table을 만들고 그 안에 {id:1}이라는 문서를 저장. 
* _id : 몽고디비자체의 자동생성 id

In [6]:
%%writefile src/ds_save_mongo_test.js
use test
db.testCollection.insert({x:1})
db.testCollection.find()
db.stats()

Writing src/ds_save_mongo_test.js


In [7]:
!mongo < src/ds_save_mongo_test.js

MongoDB shell version v3.4.3
connecting to: mongodb://127.0.0.1:27017
MongoDB server version: 3.4.3
switched to db test
WriteResult({ "nInserted" : 1 })
{ "_id" : ObjectId("58ec947a28e86d6cfd8b22cd"), "id" : 1 }
{ "_id" : ObjectId("58ec951f383df49c45323859"), "x" : 1 }
{
	"db" : "test",
	"collections" : 1,
	"views" : 0,
	"objects" : 2,
	"avgObjSize" : 33.5,
	"dataSize" : 67,
	"storageSize" : 16384,
	"numExtents" : 0,
	"indexes" : 1,
	"indexSize" : 16384,
	"ok" : 1
}
bye


## 문제 S-5: MongDB에 'id', '이름'을 저장하고, 조회하는 프로그램을 실행하기

In [8]:
%%writefile src/ds_save_mongo_5.js
use myDB
show dbs
show tables
db.myCol.insert({"Persons":[{"id":"405", "이름":"js1"},{"id":"406", "이름":"js2"}]})
db.myCol.find({ "Persons.이름": "js1" })

Writing src/ds_save_mongo_5.js


## pymongo

In [10]:
import pymongo

In [11]:
from pymongo import MongoClient

client = MongoClient() #파이썬창에서 클라이언트 생성 '>'
db=client.myDB #없으면 만들고 있으면 있는거 쓴다

#### 입력해보기

In [12]:
_id=1
_name='js'
_age=11
_country='ko'

In [13]:
db.myPyCol.insert_one({ #myPyCol이란 장표에가서 insert_one이라는 행추가를 하는것.
    "id":_id,
    "name":_name,
    "age":_age,
    "country":_country
})

In [14]:
results=db.myPyCol.find() #find()는 cursor을 만들어준다.
for r in results:
    print r['name'] #key로 검색

js


In [14]:
_name='jslim'
_age=123
db.myPyCol.update_one(
{"id":_id},
    {"$set":{
        "name":_name,
        "age":_age,
    }}
)

## 문제 6번

In [7]:
%%writefile src/ds_save_mongo_6.js
use myDB
show dbs
db.myCol.drop()
db.myCol.insert({"Persons":[{"id":"405", "이름":"js1"},{"id":"406", "이름":"js2"}]})
results=db.myCol.find()
for r in results:
    print r['Persons']
              

Overwriting src/ds_save_mongo_6.js


In [19]:
%%writefile src/ds_save_mongo_6.py
#coding:UTF-8
import pymongo
from pymongo import MongoClient

client = MongoClient()
db=client.myDB
db.myCol.drop()
db.myCol.insert_one(
   {"Persons":[{"id":"405", "이름":"js1"},{"id":"406", "이름":"js2"}]}
)
results = db.myCol.find()
for r in results:
        print r['Persons']

Overwriting src/ds_save_mongo_6.py


1. standard i/o

2. file i/o
	- json (','로 구분하는 형식) 
		csv -> delimiter(,)
        
3. db i/o 
	- mongodb


서버를 설치하게되면 먼저 설정을 해야한다. (configuartion)

* port 번호 
* -web: 80
* -ssh: 22
* -몽고db:27017
* -개인 : 1000이상으로 지정
* -tomcat : 8080

사실 www.smu.ac.kr:80 의 80이 생략되있는 것임


몽고디비 디렉터리에 있는 mongod.lock 파일은 쓰레기 파일임. 지워도 상관없음. 지워주는게 좋다.

mongod의 d는 demon을 의미한다.

pymongo는 ? 

mongo를 쓸때 shell에서 쓸 수도 있다. command창에서 직접 쓰는것.

python에서 mongo를 쓰는것은 내장형 프로그램 pymongo는 python에서 몽고를 쓰는것이다. 즉, 임베디드형태로 쓰는것이라고 할 수 있다.

열린데이터와 공공데이터포탈사이트에 자신의 key를 만들어오기 . utf-8로 받아야한다.